Move exit_if early in consumer_loop

bibstha opened the pull request:
While debugging our Kafka Consumer Jobs, I realized that `exit_if` is never called if `fetch_batches` raises an Exception, and instead it keeps running in a never terminating loop.
This PR moves `exit_if` check as the first thing we do before we fetch anything and makes sure we exit when the condition is met.

```
$ bundle exec rspec
....
Finished in 10.86 seconds (files took 0.37654 seconds to load)
97 examples, 0 failures
```


cc: @Shopify/resiliency

bibstha commented:
Part of https://github.com/Shopify/resiliency-engineering/issues/28. This will be followed up by `gem update ruby-kafka` on Shopify-core

bibstha commented on lib/kafka/consumer.rb:
```diff
@@ -277,7 +278,6 @@ def each_batch(min_bytes: 1, max_wait_time: 1, automatically_mark_as_processed:
 
           @heartbeat.send_if_necessary
 
-          @running = false if exit_if && exit_if.call
```
Hum, since this is inside the `batches.each` which I moved outside, I wonder if I should `undelete` it?
The tests do pass with this line removed. Seems like there isn't anything that would loop or take too much time inside the `batches.each` loop. Thoughts?

fw42 replied:
Agree that this should probably stay (although I would add a `break` and move it to before the `heartbeat.send_if_necessary` since I've seen the heartbeat take forever too.

fw42 commented on lib/kafka/consumer.rb:
```diff
@@ -244,6 +244,7 @@ def each_message(min_bytes: 1, max_wait_time: 1, automatically_mark_as_processed
     # @return [nil]
     def each_batch(min_bytes: 1, max_wait_time: 1, automatically_mark_as_processed: true, exit_if: nil)
       consumer_loop do
+        @running = false if exit_if && exit_if.call
```
Shouldn't you `return` or `break` here? otherwise it's still gonna execute the next `fetch_batches` even though we already know that we want to exit.

bibstha commented:
@fw42 thanks for spotting the missing `break`. I added it and also added back the line inside `each_batch`.
Since heartbeat has it's own timeout, I think we can keep it as is?

fw42 commented on spec/consumer_spec.rb:
```diff
@@ -224,6 +224,36 @@
+        th = Thread.new do
+          started_at = Time.now
+          consumer.each_batch(exit_if: -> { Time.now > (started_at + 2) } ) do |message|
+            fail "this block is never called"
```
`s/fail/flunk/`?

bibstha replied:
It's `fail` in rspec, no?

fw42 replied:
ah sorry I was thinking minitest I guess

kirs replied:
Does this test fail if you remove the fix in `consumer.rb`?

fw42 commented on spec/consumer_spec.rb:
```diff
@@ -224,6 +224,36 @@
+      it "aborts earlier if exit_if is passed" do
+        th = Thread.new do
+          started_at = Time.now
+          consumer.each_batch(exit_if: -> { Time.now > (started_at + 2) } ) do |message|
```
I think you would get the same test coverage with `exit_if: ->{ true }`, but without making the tests 2 seconds slower

fw42 commented on spec/consumer_spec.rb:
```diff
@@ -224,6 +224,36 @@
+      end
+
+      it "aborts earlier if exit_if is passed" do
+        th = Thread.new do
```
not super clear to me why this test needs a new thread? why can't this run inline?

bibstha replied:
These tests are basically copied from https://github.com/Shopify/ruby-kafka/blob/1673aed9078ed23417cf7811ed4b33d1e4e6cac7/spec/consumer_spec.rb#L200-L225 (same file, for a different case when exit_if was introduced)

The thread is required because if the source code changes and `exit_if` was removed, this test will never terminate and loop endlessly. I think @kirs added it here to avoid that.

rafaelfranca commented:
Why are we forking a gem instead of patching upstream?

rafaelfranca commented:
Ok, found the reasoning. We are way behind in the version. We are using 0.4.3. Upstream is at 0.7.x. We should make sure we are using the latest versions and that this issue is also fixed upstream. Forks only delay problems and make our lives harder in the future.

fw42 commented:
I think ideally we will stop using this gem altogether (in core, there's already plans to get rid of it). But I agree with you, Rafael 😢

kirs commented on ruby-kafka.gemspec:
```diff
@@ -37,7 +37,7 @@ Gem::Specification.new do |spec|
   spec.add_development_dependency "colored"
   spec.add_development_dependency "rspec_junit_formatter", "0.2.2"
-  spec.add_development_dependency "dogstatsd-ruby", ">= 3.0.0"
+  spec.add_development_dependency "dogstatsd-ruby", "~> 3.0"
```
Core is using `dogstatsd-ruby (4.0.0)`. Does it mean it would conflict?

bibstha replied:
I just tested it out locally and seems like `development_dependency` is not pulled in Gemfile.lock and has no impact.
This is because datadog part of this gem is not yet used in Shopify
From readme on `0-4-3-with-exit_if` branch

```
# This enables the reporter:
require "kafka/datadog"

# Default is "ruby_kafka".
Kafka::Datadog.namespace = "custom-namespace"

# Default is "127.0.0.1".
Kafka::Datadog.host = "statsd.something.com"

# Default is 8125.
Kafka::Datadog.port = 1234
```

Which we don't do in Shopify. So all good.

kirs commented on spec/consumer_spec.rb:
```diff
@@ -224,6 +224,35 @@
+          end
+        end
+
+        expect(th.join(thread_timeout)).to be_nil # join returning nil is a sign of thread not completing within timeout
```
Thanks for leaving the comment!
